app/controllers/settings/ext_cert_controller.rb

```diff
@@ -0,0 +1,260 @@
+      message = "Error with #{key_cert_map[:name]}, certificate and key must be uploaded together."
+      render_failure_event(message)
+    # Validate cert/key and verify that they match
+    else
+      cert = read_cert(key_cert_map[:cert][:cert_string])
+      key = read_key(key_cert_map[:key][:key_string])
+
+      # Check certificate valid format
+      unless cert
+        message = "Invalid #{key_cert_map[:name]} certificate, check format and try again."
+        return render_failure_event(message)
+      end
+
+      # Check key valid format
+      unless key
+        message = "Invalid #{key_cert_map[:name]} key, check format and try again."
+        return render_failure_event(message)
+      end
+
+      # Check that key matches certificate
+      unless cert.check_private_key(key)
+        message = "#{key_cert_map[:name]} Certificate/Key pair invalid.  Ensure Certificate" \
+        " and Key are matching."
+        return render_failure_event(message)
+      end
+
+      # Check that cert has valid date range
+      return false unless cert_date_check(cert)
+
+      # Moved to another task
+      # Check that hostname is in SubjectAltName of cert
+      # return false unless hostname_check(key_cert_map[:name], cert)
+
+      # Moved to another task
+      # Check the trust chain is valid
+      # return false unless trust_chain_verify(cert)
+
+      # Everything's good!
+      true
+    end
+  end
+  # rubocop:enable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+  def cert_parse(cert_string)
+    params = {}
+
+    # Check if certificate exists, assume that validation has already occured
+    if !cert_string
+      params[:Notice] = "Certificate not available, please upload a certificate"
+    else
+      cert = read_cert(cert_string)
+      unless cert
+        # :nocov:
+        # Cert has aleady been valiated when entered, an error here would require a failure
+        # in connection from velum to salt or an unintended change in the salt pillar
+        params[:Error] = "Failed to parse stored certificate, please check format and upload again"
+        return params
+        # :nocov:
+      end
+      fingerprint = cert_fingerprint(cert)
+
+      params["Subject Alternative Name".to_sym] = get_san_array(cert)
+      params["Subject Name".to_sym] = cert.subject.to_s.tr("/", " ")
+      params["Issuer Name".to_sym] = cert.issuer.to_s.tr("/", " ")
+      params["Signature Algorithm".to_sym] = cert.signature_algorithm
+      params["Not Valid Before".to_sym] = cert.not_before
+      params["Not Valid After".to_sym] = cert.not_after
+      params[fingerprint[0]] = fingerprint[1]
+    end
+    params
+  end
+
+  def cert_fingerprint(cert)
+    # Error-checking ignored, cert already validated and verified
+    fingerprint = OpenSSL::Digest::SHA256.new(cert.to_der)
+    fingerprint_string = fingerprint.to_s.scan(/../).map(&:upcase).join(":")
```

Comment:
This needs a comment saying "Add colons every 2 characters" or something similar

Also, is there no ruby string 'to_upper'? Using map(&:upcase) is a really goofy way to uppercase.

Reply:
Adding descriptive comment.  `upcase` is the way to go:  https://ruby-doc.org/core-2.2.3/String.html#method-i-upcase
I also rearranged the command to get rid of `map` and make more readable.